migration/transitioning.rst

```diff
@@ -384,9 +387,225 @@ currently logged in customer - even though this is complex analytical
+
+.. code-block:: python
+
+  from django.utils import timezone
+  from django.db import models
+
+  class Store(models.Model):
+      name = models.CharField(max_length=255)
+      url = models.URLField()
+
+  class Product(models.Model):
+      name = models.CharField(max_length=255)
+      description = models.TextField()
+      price = models.DecimalField(max_digits=6, decimal_places=2),
+      quantity = models.IntegerField()
+      store = models.ForeignKey(Store)
+
+  class Purchase(models.Model):
+      ordered_at = models.DateTimeField(default=timezone.now)
+      billing_address = models.TextField()
+      shipping_address = models.TextField()
+
+      product = models.ForeignKey(Product)
+
+The tricky thing with this pattern is that in order to find all purchases for a store, you'll have to query for all of a store's products first. This becomes a problem once you start sharding data, and in particular when you run UPDATE or DELETE queries on nested models like purchases in this example.
+
+**1. Introduce a column for the store\_id on every record that belongs to a store**
+
+In order to scale out a multi-tenant model, it's essential that you can locate all records that belong to a store quickly. The easiest way to achieve this is to simply add a :code:`store_id` column on every object that belongs to a store, and backfill your existing data to have this column set correctly.
+
+**2. Use UNIQUE constraints which include the store\_id**
+
+Unique constraints on values will present a problem in any distributed system, since it’s difficult to make sure that no two nodes accept the same unique value.
+
+In many cases, you can work around this problem by adding the store\_id to the constraint, effectively making objects unique inside a given store, but not guaranteeing this beyond that store.
+
+Let's begin by adjusting our model definitions and have Django generate a new migration for the two changes discussed.
+
+.. code-block:: python
+
+  from django.utils import timezone
+  from django.db import models
+
+  class Store(models.Model):
+    name = models.CharField(max_length=255)
+    url = models.URLField()
+
+  class Product(models.Model):
+    name = models.CharField(max_length=255)
+    description = models.TextField()
+    price = models.DecimalField(max_digits=6, decimal_places=2),
+    quantity = models.IntegerField()
+    store = models.ForeignKey(Store)
+
+    class Meta(object):                  # added
+      unique_together = ["id", "store"]  #
+
+  class Purchase(models.Model):
+    ordered_at = models.DateTimeField(default=timezone.now)
+    billing_address = models.TextField()
+    shipping_address = models.TextField()
+
+    product = models.ForeignKey(
+      Product,
+      db_constraint=False                # added
```

Comment:
Later we have a custom migration:

```sql
ALTER TABLE mtdjango_purchase
ADD CONSTRAINT mtdjango_purchase_product_fk
FOREIGN KEY (store_id, product_id)
REFERENCES mtdjango_product (store_id, id)
ON DELETE CASCADE;
```

I believe that declaring a models.ForeignKey relation in django between product and purchase allows succinct requests for a product's purchases in the ORM. I don't think that merely adding another uniqueness constraint would do that. (When the user uses your multi tenant library the query will include the tenant id as well.)